### PR TITLE
Expand tf hub to do multiple labels

### PR DESCRIPTION
**Pull request by sorensenjs**
Entries that have missing labels will not back propagate their corrections for the missing entries.

### CODE REVIEW

File: experiments/tf_trainer/common/tfrecord_input.py [hunk 1]
```diff
@@ -74,6 +74,18 @@ def _input_fn_from_file(self, filepath: str) -> types.FeatureAndLabelTensors:
...
     return batched_dataset
 
+  def _process_labels(self, features, parsed):
+    labels = {}
+    for label in self._labels:
+      label_value = parsed[label]
+      # Missing weights are negative, find them and zero those features out.
+      weight = tf.cast(tf.greater_equal(label_value, 0.0), dtype=tf.float32)
+      if self._round_labels:
+        label_value = tf.round(label_value)
```
[the remark this answers is not in the document]
Comment (sorensenjs) [reply experiments/tf_trainer/common/tfrecord_input.py]: Mask is usually a boolean (it's in there, the tf.greater_equal()) and I suppose it could be pulled out. I'm not really happy about any of this - the fact that the weight for the labels has to go on the features[] dict side instead of the labels[] dict side just seems wrong - there are no weights at inference time. But here it is
https://github.com/tensorflow/estimator/blob/75d63af79baf124c0b5bff427bd4536a7bd517f7/tensorflow_estimator/python/estimator/head/base_head.py#L479

My cloud job is failing for a mysterious reason
https://pantheon.corp.google.com/logs/viewer?project=wikidetox&minLogLevel=0&expandAll=false&timestamp=2018-11-20T22:05:17.563000000Z&customFacets=&limitCustomFacetWidth=true&dateRangeStart=2018-11-20T20:43:04.759Z&interval=PT1H&resource=ml_job&scrollTimestamp=2018-11-20T22:01:34.193660497Z&dateRangeUnbound=forwardInTime
I'm trying to run locally to see if it reveals something.

[the remark this answers is not in the document]
Comment (fprost) [reply experiments/tf_trainer/common/tfrecord_input.py]: I agree with you it is weird to put weights in the features whereas it is not part of the inference graph. However that is always what I have seen.

Not sure where the error is coming from. Your model is exported just before and the job is reported as successful. Could be any kind of warning message to be ignored?

[the remark this answers is not in the document]
Comment (sorensenjs) [reply experiments/tf_trainer/common/tfrecord_input.py]: Maybe - but why is it terminating after so few iterations?

[the remark this answers is not in the document]
Comment (fprost) [reply experiments/tf_trainer/common/tfrecord_input.py]: Is it possible that it is the end of your file? What I have noticed is that you do not have dataset.repeat(None). Which means that is is iterating only once over the file.

[the remark this answers is not in the document]
Comment (sorensenjs) [reply experiments/tf_trainer/common/tfrecord_input.py]: Investigating that now, that's true of the other code path as well, are we only making one pass?

[the remark this answers is not in the document]
Comment (fprost) [reply experiments/tf_trainer/common/tfrecord_input.py]: The tokenizer part has it: https://github.com/conversationai/conversationai-models/blob/a4f16207c25dcd5559d464457d21b24375b71cf7/experiments/tf_trainer/common/tfrecord_input.py#L134

Does batch_size*train_steps match the number of lines?

[the remark this answers is not in the document]
Comment (sorensenjs) [reply experiments/tf_trainer/common/tfrecord_input.py]: Seems like a strange place to hide the repeat
The question about the number of lines seems odd if we want to be able to train on more than one dataset.

[the remark this answers is not in the document]
Comment (fprost) [reply experiments/tf_trainer/common/tfrecord_input.py]: Sorry the question was only for debugging. I agree that the input pipeline should be independent from that.

Typically, I run an unlimited number of epochs (by applying repeat(None)). the parameter `train_steps` then regulates how long we train for.

let me know if that makes sense

File: experiments/tf_trainer/tf_hub_classifier/model.py [hunk 2]
```diff
@@ -24,7 +24,7 @@
 # This would normally just be a multi_integer, but we use string due to
 # constraints with ML Engine hyperparameter tuning.
 tf.app.flags.DEFINE_string(
-    'dense_units', '512,128,64',
+    'dense_units', '1024,1024,512',
```
Comment (iislucas) [inline experiments/tf_trainer/tf_hub_classifier/model.py]: Why the same number twice?

File: experiments/tf_trainer/common/tfrecord_input.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (sorensenjs) [reply experiments/tf_trainer/common/tfrecord_input.py]: Resolving this for potential future follow up.

File: experiments/tf_trainer/tf_hub_classifier/model.py (hunk 2 above)
Comment (sorensenjs) [reply experiments/tf_trainer/tf_hub_classifier/model.py]: Multiple layers, all the same size.

Comment (iislucas) [reply experiments/tf_trainer/tf_hub_classifier/model.py]: Ah, thanks. Then maybe add to the description to say: "number of hidden units in in the dense layers." ?
